test/test_main.py

```diff
@@ -0,0 +1,276 @@
+                'CHARM_CONFIG': charm_config,
+            })
+        return self._read_and_clear_state()
+
+    def test_event_reemitted(self):
+        # Test case stats.
+        stats = {
+            'count_called': 0,
+            'count_deferred': 0,
+            'event_type_names_observed': list(),
+        }
+
+        def update_stats(state, stats):
+            state = self._read_and_clear_state()
+            stats['count_called'] += state.get(
+                'ConfigChangedEvent_count_called', 0)
+            stats['count_deferred'] += state.get(
+                'ConfigChangedEvent_count_deferred', 0)
+            stats['event_type_names_observed'].extend(
+                state.get('event_type_names_observed', []))
+
+        # base64 encoding is used to avoid null bytes.
+        charm_config_defer = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+            'ConfigChangedEvent_defer': True
+        }))
+
+        # First run "install" to make sure all hooks are set up.
+        update_stats(self._simulate_event('install',
+                                          charm_config_defer), stats)
+
+        self.assertEqual(stats['count_called'], 0)
+        self.assertEqual(stats['count_deferred'], 0)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent'])
+
+        update_stats(self._simulate_event('config-changed',
+                                          charm_config_defer), stats)
+
+        self.assertEqual(stats['count_called'], 1)
+        self.assertEqual(stats['count_deferred'], 1)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent', 'ConfigChangedEvent'])
+
+        charm_config_dont_defer = base64.b64encode(pickle.dumps({
+            'STATE_FILE': self._state_file,
+            'ConfigChangedEvent_defer': False
+        }))
+
+        # Re-emit should pick the deferred config-changed.
+        update_stats(self._simulate_event('update-status',
+                                          charm_config_dont_defer), stats)
+
+        self.assertEqual(stats['count_called'], 2)
+        self.assertEqual(stats['count_deferred'], 1)
+        self.assertEqual(stats['event_type_names_observed'],
+                         ['InstallEvent', 'ConfigChangedEvent',
+                          'ConfigChangedEvent', 'UpdateStatusEvent'])
+
+    def test_all_events_handled(self):
+        # Infer supported events from the charm class and CharmEvents.
+        charm_events = {}
+        charm_events.update({event_name: instance for event_name, instance
+                             in CharmEvents.__dict__.items()
+                             if type(instance) == Event})
+        charm_events.update({event_name: instance for event_name, instance
+                             in type(charm.Charm.on).__dict__.items()
+                             if type(instance) == Event})
```

Comment:
Same comments here to the one in the sample charm.

Reply:
See below.